# Holman Data Science exercise challenge

- Perform this analysis on the most recent 3 years of data
- Illustrate the Mean, Standard Deviation, and Median of the data set.
- Find which prices are outliers and how you determined them to be so.
- Determine what should be the expected Ford F150 tire repair part price.

## INSTALL LIBRARIES

In [ ]:
import sys
!{sys.executable} -m pip install pyarrow

In [1]:
import pandas as pd

## LOAD AND PREPARE DATA

In [2]:
df = pd.read_parquet('data/f150-tire-repairs.parquet')

In [3]:
df

part_id    cost                date        id  \
236       17001001  152.00 2017-06-15 14:59:32  67108851   
437       17001001  291.45 2023-02-08 11:03:58  98417990   
700       17001001  217.39 2021-05-19 15:31:44  89322574   
708       17001001  241.76 2022-11-16 15:53:45  97280532   
710       17001001  234.01 2022-02-28 11:10:51  93414648   
...            ...     ...                 ...       ...   
26194485  17001001  182.00 2021-11-16 17:31:44  91938811   
26194513  17001001  171.00 2021-01-22 09:19:14  87543421   
26194861  17001001  187.00 2021-05-04 09:39:24  89003083   
26196818  17001001  186.00 2019-10-21 22:33:01  80578102   
26196855  17001001  272.00 2021-04-21 07:27:39  88823219   

               repair_description  make  model  year  
236       TIRE, RADIAL  LUG TREAD  FORD  F-150  2015  
437       TIRE, RADIAL  LUG TREAD  FORD  F-150  2022  
700       TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
708       TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
710       TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
...                           ...   ...    ...   ...  
26194485  TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
26194513  TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
26194861  TIRE, RADIAL  LUG TREAD  FORD  F-150  2015  
26196818  TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
26196855  TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  

[29445 rows x 8 columns]

In [29]:
# Are all the repair_description values tire repairs?  Cannot verify with information provided
df.repair_description.unique()

array(['TIRE, RADIAL  LUG TREAD', 'TRANSFORCE AT2 LT2 2',
       'TRANSFORCE HT2 LT265/70R17 E 1', 'ROTATE', '245/70R17 GENERAL GR',
       'MOUNT & DISMOUNT TIR', '245/70R17 GENERAL BR',
       'LT265/70R17 COOPER', 'TRANSFORCE AT2     2',
       'DESTINATION M/T2 LT265/70R17 E', 'DUELER A/T REVO 3 LT265/70R17',
       'PATCH, WHEEL WEIGHTS', 'TIRE,2457017 LTX MS2',
       'LT265/70R17112/109SA', 'LT265/70R17 ST MAXX',
       '245/70R17 MICH LTX M', 'FOUR TIRE ROTATION',
       'LT245/70R17 E / DISC', 'LT245/75R17 HANKOOK',
       'LT245/75R17 COOPER D', 'DESTINATION A/T2   2',
       '265/60R18 109T SR WR', '235 70R17KUMHO XL',
       'TRANSFORCE AT2 LT265/70R17 E 1', '245/70R17 110T ADVTASPTLT GO',
       '265/70R17 OPN CTRY', 'Miscellaneous Tire Service, Pa',
       'TRANSFORCE AT2 LT245/70R17 E 1', 'TRANSFORCE AT2 LT245/75R17 E 1',
       'MISC TIRE', 'MISC - MISCELLANEOUS', '245/70R17 FSTONE DES',
       'LT245/75/R17 FIRESTO', 'LT265/70R17112/109SATTAKO2LRC',
       'LT265/70R17

In [4]:
df['cost'].describe()

count    29445.000000
mean       222.388635
std         54.159489
min          1.000000
25%        182.970000
50%        219.000000
75%        255.200000
max       1216.880000
Name: cost, dtype: float64

### Create dataset for the most recent 3 years

I am interpretting "Perform this analysis on the most recent 3 years of data" to mean three years back from today 

In [6]:
from datetime import datetime

# End date is three years ago
end_date = datetime.strptime('5/17/20 7:37:11', '%m/%d/%y %H:%M:%S')

three_yr_df = df.loc[(df['date'] >= end_date)]
three_yr_df.sort_values(by=['date'])

part_id    cost                date        id  \
22603571  17001001  149.99 2020-05-18 09:43:41  83709063   
9794827   17001001  196.99 2020-05-18 09:49:11  83709147   
24223746  17001001  235.00 2020-05-18 10:49:33  83710551   
9330524   17001001  168.99 2020-05-18 10:52:02  83710591   
6225214   17001001  239.95 2020-05-18 11:18:36  83711223   
...            ...     ...                 ...       ...   
20625359  17001001  208.99 2023-04-28 15:48:38  99701298   
9098727   17001001  304.67 2023-04-28 17:43:41  99703604   
7372935   17001001  200.90 2023-04-28 18:23:19  99704138   
14149801  17001001  176.47 2023-04-29 10:57:35  99709483   
11247496  17001001  241.66 2023-04-30 09:59:23  99713264   

               repair_description  make  model  year  
22603571  TIRE, RADIAL  LUG TREAD  FORD  F-150  2016  
9794827   TIRE, RADIAL  LUG TREAD  FORD  F-150  2015  
24223746  TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
9330524   TIRE, RADIAL  LUG TREAD  FORD  F-150  2013  
6225214   TIRE, RADIAL  LUG TREAD  FORD  F-150  2015  
...                           ...   ...    ...   ...  
20625359  TIRE, RADIAL  LUG TREAD  FORD  F-150  2021  
9098727   TIRE, RADIAL  LUG TREAD  FORD  F-150  2010  
7372935   TIRE, RADIAL  LUG TREAD  FORD  F-150  2019  
14149801  TIRE, RADIAL  LUG TREAD  FORD  F-150  2014  
11247496  TIRE, RADIAL  LUG TREAD  FORD  F-150  2017  

[18855 rows x 8 columns]

In [8]:
# The most recent date was actually 4/30 so I need to adjust to get a full three years
end_date = datetime.strptime('4/30/20 9:59:23', '%m/%d/%y %H:%M:%S')

three_yr_df = df.loc[(df['date'] >= end_date)]
three_yr_df.sort_values(by=['date'])

part_id    cost                date        id  \
16726183  17001001  245.19 2020-04-30 10:55:17  83473092   
8833068   17001001  158.38 2020-04-30 11:04:14  83473390   
7349197   17001001  195.00 2020-04-30 11:05:12  83473413   
12374356  17001001  195.00 2020-04-30 11:58:43  83474902   
25172864  17001001  137.95 2020-04-30 12:50:02  83476126   
...            ...     ...                 ...       ...   
20625359  17001001  208.99 2023-04-28 15:48:38  99701298   
9098727   17001001  304.67 2023-04-28 17:43:41  99703604   
7372935   17001001  200.90 2023-04-28 18:23:19  99704138   
14149801  17001001  176.47 2023-04-29 10:57:35  99709483   
11247496  17001001  241.66 2023-04-30 09:59:23  99713264   

               repair_description  make  model  year  
16726183  TIRE, RADIAL  LUG TREAD  FORD  F-150  2011  
8833068   TIRE, RADIAL  LUG TREAD  FORD  F-150  2017  
7349197   TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
12374356  TIRE, RADIAL  LUG TREAD  FORD  F-150  2010  
25172864  TIRE, RADIAL  LUG TREAD  FORD  F-150  2016  
...                           ...   ...    ...   ...  
20625359  TIRE, RADIAL  LUG TREAD  FORD  F-150  2021  
9098727   TIRE, RADIAL  LUG TREAD  FORD  F-150  2010  
7372935   TIRE, RADIAL  LUG TREAD  FORD  F-150  2019  
14149801  TIRE, RADIAL  LUG TREAD  FORD  F-150  2014  
11247496  TIRE, RADIAL  LUG TREAD  FORD  F-150  2017  

[19046 rows x 8 columns]

## PERFORM ANALYSIS

### Illustrate the Mean, Standard Deviation, and Median of the data set.

In [9]:
three_yr_df['cost'].mean() # The mean

231.33586842381604

In [10]:
three_yr_df['cost'].std() # The standard deviation

55.0341497865612

In [11]:
three_yr_df['cost'].median() # The median

226.67

In [12]:
three_yr_df['cost'].describe()

count    19046.000000
mean       231.335868
std         55.034150
min          1.000000
25%        192.000000
50%        226.670000
75%        265.000000
max       1216.880000
Name: cost, dtype: float64

### Find the outliers

- My approach will be to use the Z-score and IQR methods to find the outliers
- I am not choosing which set of outliers are the 'correct' ones, per se, but will use both sets to find the final price
- If a choice is required, I'll go with the IQR set

#### Z-score

In [14]:
upper_limit = three_yr_df['cost'].mean() + 3* df['cost'].std()
lower_limit = three_yr_df['cost'].mean() - 3* df['cost'].std()
print(upper_limit)
print(lower_limit)

393.81433638475005
68.857400462882


In [15]:
outlier_df = three_yr_df.loc[(three_yr_df['cost'] <= lower_limit) | (three_yr_df['cost'] >= upper_limit)]
outlier_df.sort_values(by=['cost'])

part_id     cost                date        id  \
20923996  17001001     1.00 2022-03-25 13:46:18  93820847   
13443980  17001001     1.00 2022-03-24 16:42:21  93807125   
24783259  17001001     2.53 2022-11-04 11:15:40  97111914   
17842113  17001001     3.00 2021-06-24 14:22:04  89823438   
22715321  17001001     4.59 2021-03-26 12:46:53  88446294   
...            ...      ...                 ...       ...   
20573543  17001001   575.00 2022-11-18 11:06:02  97309917   
5125386   17001001   610.82 2023-01-17 10:10:32  98081346   
10451158  17001001   686.84 2022-02-28 17:12:59  93424908   
12915899  17001001   825.00 2022-12-07 12:29:31  97550160   
23174884  17001001  1216.88 2022-11-29 15:28:50  97434269   

               repair_description  make  model  year  
20923996  TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
13443980  TIRE, RADIAL  LUG TREAD  FORD  F-150  2013  
24783259  TIRE, RADIAL  LUG TREAD  FORD  F-150  2015  
17842113  TIRE, RADIAL  LUG TREAD  FORD  F-150  2015  
22715321  TIRE, RADIAL  LUG TREAD  FORD  F-150  2017  
...                           ...   ...    ...   ...  
20573543  TIRE, RADIAL  LUG TREAD  FORD  F-150  2021  
5125386   TIRE, RADIAL  LUG TREAD  FORD  F-150  2021  
10451158  TIRE, RADIAL  LUG TREAD  FORD  F-150  2016  
12915899  TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
23174884  TIRE, RADIAL  LUG TREAD  FORD  F-150  2013  

[147 rows x 8 columns]

#### Interquartile Range (IQR)
Now I'll compare using IQR as a sanity check

In [16]:
IQR = three_yr_df['cost'].quantile(0.75) - three_yr_df['cost'].quantile(0.25)

iqr_upper_limit = three_yr_df['cost'].quantile(0.75) + (IQR * 1.5)
iqr_lower_limit = three_yr_df['cost'].quantile(0.25) - (IQR * 1.5)
print(iqr_upper_limit)
print(iqr_lower_limit)

374.5
82.5


In [17]:
iqr_outlier_df = three_yr_df.loc[(three_yr_df['cost'] <= iqr_lower_limit) | (three_yr_df['cost'] >= iqr_upper_limit)]
iqr_outlier_df.sort_values(by=['cost'])

part_id     cost                date        id  \
13443980  17001001     1.00 2022-03-24 16:42:21  93807125   
20923996  17001001     1.00 2022-03-25 13:46:18  93820847   
24783259  17001001     2.53 2022-11-04 11:15:40  97111914   
17842113  17001001     3.00 2021-06-24 14:22:04  89823438   
22715321  17001001     4.59 2021-03-26 12:46:53  88446294   
...            ...      ...                 ...       ...   
20573543  17001001   575.00 2022-11-18 11:06:02  97309917   
5125386   17001001   610.82 2023-01-17 10:10:32  98081346   
10451158  17001001   686.84 2022-02-28 17:12:59  93424908   
12915899  17001001   825.00 2022-12-07 12:29:31  97550160   
23174884  17001001  1216.88 2022-11-29 15:28:50  97434269   

               repair_description  make  model  year  
13443980  TIRE, RADIAL  LUG TREAD  FORD  F-150  2013  
20923996  TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
24783259  TIRE, RADIAL  LUG TREAD  FORD  F-150  2015  
17842113  TIRE, RADIAL  LUG TREAD  FORD  F-150  2015  
22715321  TIRE, RADIAL  LUG TREAD  FORD  F-150  2017  
...                           ...   ...    ...   ...  
20573543  TIRE, RADIAL  LUG TREAD  FORD  F-150  2021  
5125386   TIRE, RADIAL  LUG TREAD  FORD  F-150  2021  
10451158  TIRE, RADIAL  LUG TREAD  FORD  F-150  2016  
12915899  TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
23174884  TIRE, RADIAL  LUG TREAD  FORD  F-150  2013  

[272 rows x 8 columns]

#### IQR identifies more outliers, now I'll see how it affects  thefinal outcome

### Determine what should be the expected Ford F150 tire repair part price

- The Z-score and IQR outlier results do not match very well
- Perform central tendency analysis on both, hopefully those results will be close 
- 
- The expected repair part price should be the central tendency of the price, either median or mean
- Once the outliers are removed, I think the mean would be more appropriate as median is more useful if there are outliers
- I'll find the mean for each set (Z-score and IQR) and see if the mean is relatively close

#### z_clean_df is the three year dataframe without Z-score outliers

In [23]:
z_clean_df = three_yr_df.loc[(three_yr_df['cost'] > lower_limit) & (three_yr_df['cost'] < upper_limit)]
z_clean_df.sort_values(by=['cost'])

# upper_limit: 393.81433638475005
# lower_limit: 68.857400462882

part_id    cost                date        id  \
4923449   17001001   69.02 2022-08-16 13:50:57  95968233   
26091327  17001001   70.20 2021-04-21 11:47:32  88829891   
19275593  17001001   75.00 2022-04-05 13:12:57  93969653   
4068042   17001001   75.00 2021-07-22 09:33:32  90181537   
5820795   17001001   75.79 2021-08-25 11:16:49  90654937   
...            ...     ...                 ...       ...   
6706446   17001001  391.71 2022-07-12 13:35:53  95474271   
25358025  17001001  392.00 2023-03-03 11:38:24  98755931   
19680027  17001001  392.64 2022-08-15 10:19:47  95942949   
4911712   17001001  393.32 2023-01-30 17:57:09  98286968   
10630003  17001001  393.60 2022-04-27 18:33:07  94290703   

               repair_description  make  model  year  
4923449   TIRE, RADIAL  LUG TREAD  FORD  F-150  2022  
26091327  TIRE, RADIAL  LUG TREAD  FORD  F-150  2015  
19275593  TIRE, RADIAL  LUG TREAD  FORD  F-150  2008  
4068042   TIRE, RADIAL  LUG TREAD  FORD  F-150  2017  
5820795   TIRE, RADIAL  LUG TREAD  FORD  F-150  2016  
...                           ...   ...    ...   ...  
6706446   TIRE, RADIAL  LUG TREAD  FORD  F-150  2013  
25358025  TIRE, RADIAL  LUG TREAD  FORD  F-150  2017  
19680027  TIRE, RADIAL  LUG TREAD  FORD  F-150  2019  
4911712   TIRE, RADIAL  LUG TREAD  FORD  F-150  2021  
10630003  TIRE, RADIAL  LUG TREAD  FORD  F-150  2019  

[18899 rows x 8 columns]

In [24]:
z_clean_df['cost'].describe()

count    18899.000000
mean       230.404448
std         51.499775
min         69.020000
25%        192.000000
50%        226.000000
75%        264.170000
max        393.600000
Name: cost, dtype: float64

#### iqr_clean_df is the three year dataframe without IQR outliers

In [25]:
iqr_clean_df = three_yr_df.loc[(three_yr_df['cost'] > iqr_lower_limit) & (three_yr_df['cost'] < iqr_upper_limit)]
iqr_clean_df.sort_values(by=['cost'])

# iqr_upper_limt: 374.5
# iqr_lower_limit: 82.5

part_id    cost                date        id  \
9599123   17001001   83.62 2020-09-16 11:59:22  85357655   
22071     17001001   84.99 2020-08-14 17:19:40  84932517   
9157588   17001001   87.08 2023-04-05 14:43:06  99342736   
10630259  17001001   88.00 2021-07-01 10:57:50  89911455   
15992759  17001001   90.99 2021-08-10 11:54:58  90440030   
...            ...     ...                 ...       ...   
19615670  17001001  373.71 2023-01-27 11:12:17  98251242   
5604305   17001001  373.93 2022-11-16 18:21:02  97283193   
5882124   17001001  373.99 2022-10-21 09:37:40  96894341   
17067537  17001001  374.00 2021-08-16 12:40:34  90516578   
5882860   17001001  374.32 2023-04-11 10:26:23  99405717   

               repair_description  make  model  year  
9599123   TIRE, RADIAL  LUG TREAD  FORD  F-150  2007  
22071     TIRE, RADIAL  LUG TREAD  FORD  F-150  2019  
9157588   TIRE, RADIAL  LUG TREAD  FORD  F-150  2019  
10630259  TIRE, RADIAL  LUG TREAD  FORD  F-150  1996  
15992759  TIRE, RADIAL  LUG TREAD  FORD  F-150  2016  
...                           ...   ...    ...   ...  
19615670  TIRE, RADIAL  LUG TREAD  FORD  F-150  2020  
5604305   TIRE, RADIAL  LUG TREAD  FORD  F-150  2017  
5882124   TIRE, RADIAL  LUG TREAD  FORD  F-150  2020  
17067537  TIRE, RADIAL  LUG TREAD  FORD  F-150  2018  
5882860   TIRE, RADIAL  LUG TREAD  FORD  F-150  2019  

[18774 rows x 8 columns]

In [26]:
iqr_clean_df['cost'].describe()

count    18774.000000
mean       229.531584
std         50.122022
min         83.620000
25%        192.000000
50%        226.000000
75%        263.000000
max        374.320000
Name: cost, dtype: float64

#### Z-score mean 230.404448   ------   IQR mean 229.531584
There's no "right" answer per se, but I'll just take the average as a personal predilection

In [22]:
print((230.404448 + 229.531584) / 2)

229.968016


## Final price $229.97
The mean of the three year dataframe without Z-score outliers was quite close to the mean of the three year dataframe without IQR outliers